src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/ReadOnlyListExtensions.cs

```diff
@@ -36,6 +52,49 @@ static ImmutableArray<TResult> BuildResult(IReadOnlyList<T> items, Func<T, TResu
         }
     }
 
+    /// <summary>
+    ///  Projects each element of an <see cref="IReadOnlyList{T}"/> into a new form by incorporating the element's index.
+    /// </summary>
+    /// <typeparam name="T">The type of the elements in <paramref name="list"/>.</typeparam>
+    /// <typeparam name="TResult">The type of the value returned by <paramref name="selector"/>.</typeparam>
+    /// <param name="list">An <see cref="IReadOnlyList{T}"/> of values to invoke a transform function on.</param>
+    /// <param name="selector">
+    ///  A transform function to apply to each element; the second parameter of the function represents the index of the element.
+    /// </param>
+    /// <returns>
+    ///  Returns a new <see cref="ImmutableArray{T}"/> whose elements are the result of invoking the transform function
+    ///  on each element of <paramref name="list"/>.
+    /// </returns>
+    public static ImmutableArray<TResult> SelectAsArray<T, TResult>(this IReadOnlyList<T> list, Func<T, int, TResult> selector)
+    {
+        if (list is ImmutableArray<T> array)
+        {
+            return ImmutableArrayExtensions.SelectAsArray(array, selector);
+        }
+
+        return list switch
+        {
+            [] => [],
+            [var item] => [selector(item, 0)],
```

Comment:
>[var item] => [selector(item, 0)], [](http://example.com/codeflow?start=12&length=34)

Out of curiosity, do you think there is a benefit to special casing the 1-4 cases instead of just falling through to the BuildResult behavior?

Reply:
You make a great point. I'll update these to just create the result array up front.